### PR TITLE
Addons CRD for optionally installing Addons

### PR DESCRIPTION
**Pull request by nikhil-thomas**
# Changes

This patch is an refactored implementation of PR from Akihiko (Aki) Kuroda @akihikokuroda (taken up with his permission)
https://github.com/openshift/tektoncd-pipeline-operator/pull/23

The goal of this patch is to provide a CRD `addon.operator.tekton.dev` using which we can add addons along with tekton-pipelines

sample CR
```
apiVersion: operator.tekton.dev/v1alpha1
kind: Addon
metadata:
  name: dashboard
spec:
  version: v0.1.0
```

Add new CRD addon.operator.tekton.dev
Add a controller for reconciling addons
Add Addon CRD reconcile logic

Add Tekton-Dashboard v0.1.1 payload
Add Tekton-Dashboard v0.1.0 payload

Add E2E tests for Addon feature
Add E2E tests for testing addon installation adn addon deletion
Add make target for running e2e tests (can be used during local development)
Updates Readme

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [X] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [X] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [X] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/operator/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

```
- API changes
  - Adds `addons.operator.tekton.dev` api version `v1alpha1
  - Using addons.operator.tekton.dev Tekton Addons can be installed using this operator
  - Currently, the operator packages Tekton-Dashboard v0.1.0 and v0.1.1

```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,4 @@
+.PHONY: local-dev
+local-dev:
+	operator-sdk up local --namespace ""
```
Comment (sthaha) [inline Makefile]: how about: 
```
operator-sdk up local --namespace "" --operator-flags '--zap-encoder=console'
```

File: deploy/crds/operator_v1alpha1_addon_crd.yaml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+metadata:
+  name: addons.operator.tekton.dev
+spec:
+  group: operator.tekton.dev
+  names:
+    kind: Addon
+    listKind: AddonList
+    plural: addons
+    singular: addon
+  scope: Namespaced
```
Comment (sthaha) [inline deploy/crds/operator_v1alpha1_addon_crd.yaml]: cluster - right?

File: deploy/resources/addons/dashboard/v0.1.0/gcr-tekton-dashboard.yaml [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+subjects:
+  - kind: ServiceAccount
+    name: tekton-dashboard
+    namespace: tekton-pipelines
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: tekton-dashboard-minimal
+---
+# ------------------- Dashboard Deployment ------------------- #
```
Comment (sthaha) [inline deploy/resources/addons/dashboard/v0.1.0/gcr-tekton-dashboard.yaml]: will need `namespace` manifest before `namespaced` resources

File: pkg/controller/addon/addon_controller.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
```
Comment (sthaha) [inline pkg/controller/addon/addon_controller.go]: a better pattern I have seen is to just `return reconsile.Result{}, ignoreNotFound(err)` , so the whole block will reduce to 

```go
if err != nil {
  return reconsile.Result{}, ignoreNotFound(err)
}
```

**Comment by vincent-pli** [comment]
```apiVersion: operator.tekton.dev/v1alpha1
kind: Addon
metadata:
  name: my-addon
spec:
  version: v0.x.x
```
Could you describe the design more? 
Seems there is only `version` in the spec, what's it means? if I want to install addon: `trigger` but not `dashboard`, how could i do?

File: deploy/crds/operator_v1alpha1_addon_crd.yaml (hunk 2 above)
Comment (nikhil-thomas) [reply deploy/crds/operator_v1alpha1_addon_crd.yaml]: yes, it should be clusterwide

File: pkg/controller/addon/addon_controller.go (hunk 4 above)
Comment (nikhil-thomas) [reply pkg/controller/addon/addon_controller.go]: :+1: 
The controller definitely needs some refactoring and beautification

**Comment by nikhil-thomas** [comment]
> ```
> kind: Addon
> metadata:
>   name: my-addon
> spec:
>   version: v0.x.x
> ```
> 
> Could you describe the design more?
> Seems there is only `version` in the spec, what's it means? if I want to install addon: `trigger` but not `dashboard`, how could i do?

`metadata.name` will be the name of the addon
Each instance of Addon CRD will represent 1 Addon item

File: deploy/resources/addons/dashboard/v0.1.0/gcr-tekton-dashboard.yaml (hunk 3 above)
Comment (sthaha) [reply deploy/resources/addons/dashboard/v0.1.0/gcr-tekton-dashboard.yaml]: hmm .. where does this get installed thou? the same `namespace` as the operator?

File: pkg/controller/addon/addon_controller.go [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+	// TODO: targetNamspace should be decided from current tekton pipelines installation
+	// if there is no pipeline installation then the default namespace 'tekton-pipelines'
+	// can be used
+	targetNamespace := "tekton-pipelines"
+	tfs := []mf.Transformer{
+		mf.InjectOwner(res),
+		mf.InjectNamespace(targetNamespace),
+	}
+
```
Comment (akihikokuroda) [inline pkg/controller/addon/addon_controller.go]: The owner of the addon is the instance of the addon CRD.  How is the addon deleted when the pipeline is deleted?

**Comment by akihikokuroda** [comment]
I have some updates in the deploy/role.yaml file https://github.com/akihikokuroda/operator/blob/addon/deploy/role.yaml
Please take a look.

File: pkg/controller/addon/addon_controller.go (hunk 5 above)
Comment (nikhil-thomas) [reply pkg/controller/addon/addon_controller.go]: i have updated the reconcile logic today. I am setting the pipeline installation as owner
the code is functional, but needs some refactoring.

**Comment by nikhil-thomas** [comment]
> I have some updates in the deploy/role.yaml file https://github.com/akihikokuroda/operator/blob/addon/deploy/role.yaml
> Please take a look.

Thanks for the heads up

File: pkg/controller/addon/addon_controller.go [hunk 6]
```diff
@@ -140,10 +145,16 @@ func (r *ReconcileAddon) reconcileAddon(req reconcile.Request, res *op.Addon) (r
...
+	pipeline, err := r.getPipelineRes()
+	if err != nil {
+		_ = r.updateStatus(res, op.AddonCondition{
+			Code:    op.ErrorStatus,
+			Details: err.Error(),
+			Version: res.Spec.Version,
+		})
+		return reconcile.Result{}, err
+	}
+	targetNamespace := pipeline.Namespace
```
Comment (vincent-pli) [inline pkg/controller/addon/addon_controller.go]: `pipeline.Namespace`
Should be `pipeline.Spec.TargetNamespace` I  guess

File: pkg/controller/addon/addon_controller.go [hunk 7]
```diff
@@ -140,10 +145,16 @@ func (r *ReconcileAddon) reconcileAddon(req reconcile.Request, res *op.Addon) (r
...
 	tfs := []mf.Transformer{
 		mf.InjectOwner(res),
 		mf.InjectNamespace(targetNamespace),
```
Comment (vincent-pli) [inline pkg/controller/addon/addon_controller.go]: Install `addon` to the same namespace of `pipeline`, so when CR `Config` deleted, does the namespace deleted?
What's the relationship between `Config` and `Addon`, what happened if `Config` deleted. 

hmm, does anyone think `Config` is a little confuse, could we make the name more make sense? like `TektonPipeline`

Comment (vincent-pli) [reply pkg/controller/addon/addon_controller.go]: Seems `Addon` set owner to `Config`...

File: deploy/resources/addons/dashboard/v0.1.0/gcr-tekton-dashboard.yaml (hunk 3 above)
Comment (nikhil-thomas) [reply deploy/resources/addons/dashboard/v0.1.0/gcr-tekton-dashboard.yaml]: it will get the namespace from pipeline-installation

File: pkg/controller/addon/addon_controller.go (hunk 7 above)
Comment (nikhil-thomas) [reply pkg/controller/addon/addon_controller.go]: yes, addon owner is set to config.
so when pipeline is deleted, addons will be deleted.

> hmm, does anyone think Config is a little confuse, could we make the name more make sense? like TektonPipeline

we started with the name `install.operator.tekton.dev` instead of `config.operator.tekton.dev` :D
`config` is an upgrade compared to `install`
but, sure this is something we can discuss/decide :)

File: pkg/apis/operator/v1alpha1/addon_types.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
```
Comment (sthaha) [inline pkg/apis/operator/v1alpha1/addon_types.go]: these comments can be removed.

File: pkg/apis/operator/v1alpha1/addon_types.go [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// AddonSpec defines the desired state of Addon
+// +k8s:openapi-gen=true
+type AddonSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
```
Comment (sthaha) [inline pkg/apis/operator/v1alpha1/addon_types.go]: same ...

File: pkg/controller/addon/addon_controller.go [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	"time"
+)
+
+var (
+	ctrlLog = logf.Log.WithName("ctrl").WithName("addon")
+	ErrPipelineNotReady = xerrors.Errorf("tekton-pipelines not ready")
```
Comment (sthaha) [inline pkg/controller/addon/addon_controller.go]: Why are we exporting this?

File: pkg/controller/addon/addon_controller.go [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+	ErrAddonVersionUnresolved = xerrors.Errorf("tekton-pipelines not ready")
+)
+
+const (
+	DefaultTargetNs = "tekton-pipelines"
```
Comment (sthaha) [inline pkg/controller/addon/addon_controller.go]: these can be shared between controllers, so should these be moved to a separate package?

File: pkg/controller/addon/addon_controller.go (hunk 10 above)
Comment (nikhil-thomas) [reply pkg/controller/addon/addon_controller.go]: we are not using it outside this package, at present. So im making it unexported now. In the future if we want to use the error in tests we will have to export it.

File: pkg/controller/addon/addon_controller.go (hunk 11 above)
Comment (nikhil-thomas) [reply pkg/controller/addon/addon_controller.go]: that is better, done

**Comment by nikhil-thomas** [comment]
@vdemeester @akihikokuroda

File: deploy/resources/addons/dashboard/v0.1.0/gcr-tekton-dashboard.yaml [hunk 12]
```diff
@@ -0,0 +1,215 @@
+# ------------------- Extension Resource Definition ------------------- #
```
Comment (sthaha) [inline deploy/resources/addons/dashboard/v0.1.0/gcr-tekton-dashboard.yaml]: should this be removed?

Comment (sthaha) [reply deploy/resources/addons/dashboard/v0.1.0/gcr-tekton-dashboard.yaml]: v0.1.0

File: deploy/role.yaml [hunk 13]
```diff
@@ -1,6 +1,7 @@
 apiVersion: rbac.authorization.k8s.io/v1
 kind: ClusterRole
 metadata:
+  creationTimestamp: null
```
Comment (sthaha) [inline deploy/role.yaml]: can these be deleted?

File: deploy/role.yaml [hunk 14]
```diff
@@ -157,3 +159,16 @@ rules:
   - '*'
   verbs:
   - '*'
+- apiGroups:
```
Comment (sthaha) [inline deploy/role.yaml]: why is this needed?

File: deploy/role.yaml [hunk 15]
```diff
@@ -157,3 +159,16 @@ rules:
...
+  - dashboard.tekton.dev
+  resources:
+  - '*'
+  - addons
+  verbs:
+  - '*'
+- apiGroups:
```
Comment (sthaha) [inline deploy/role.yaml]: are these relevant? openshift.io?

File: pkg/controller/addon/addon_controller.go [hunk 16]
```diff
@@ -0,0 +1,358 @@
...
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+	addons map[string]mf.Manifest
+}
+
+// Reconcile reads that state of the cluster for a Addon object and makes changes based on the state read
+// and what is in the Addon.Spec
+func (r *ReconcileAddon) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log := requestLogger(req, "addon reconcile")
```
Comment (sthaha) [inline pkg/controller/addon/addon_controller.go]: nit: reconcile should do .. as the logger already has `ctrl.addon`

File: pkg/controller/addon/addon_controller.go [hunk 17]
```diff
@@ -0,0 +1,358 @@
...
+	//reqLogger := log.WithValues("Request.Namespace", req.Namespace, "Request.Name", req.Name)
```
Comment (sthaha) [inline pkg/controller/addon/addon_controller.go]: can we remove this?

File: pkg/controller/addon/addon_controller.go [hunk 18]
```diff
@@ -0,0 +1,358 @@
...
+		Namespace: res.Namespace,
+		Name:      res.Name,
+	}
+	return r.client.Get(context.TODO(), objKey, res)
+}
+
+func (r *ReconcileAddon) getPipelineRes() (*op.Config, error) {
+	res := &op.Config{}
+	namespacedName := types.NamespacedName{
+		Namespace: "",
```
Comment (sthaha) [inline pkg/controller/addon/addon_controller.go]: this can be removed - right? Default is ""

File: pkg/controller/addon/addon_controller.go [hunk 19]
```diff
@@ -0,0 +1,358 @@
...
+	log := requestLogger(req, "addon install")
+
+	err := r.updateStatus(res, op.AddonCondition{Code: op.InstallingStatus, Version: res.Spec.Version})
+	if err != nil {
+		log.Error(err, "failed to set status")
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	//find the valid clusterwide tekton-pipeline installation
+	piplnRes, err := r.pipelineReady()
```
Comment (sthaha) [inline pkg/controller/addon/addon_controller.go]: how about below to improve readabilty? 

```go
config, err := r.getClusterConfig()
if err {
...
}
if err := r.validateClusterConfig(config); err != nil {
   .. 
}
```

File: test/helpers/helpers.go [hunk 20]
```diff
@@ -67,7 +68,7 @@ func WaitForClusterCR(t *testing.T, name string) *op.Config {
 	})
 
 	AssertNoError(t, err)
-	return cr
+	//return cr
```
Comment (sthaha) [inline test/helpers/helpers.go]: delete: commented code ?

File: test/testgroups/addon_crd.go [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+	ctx := test.NewTestCtx(t)
+	defer ctx.Cleanup()
+
+	t.Log("deploying operaot for addomn")
+	err := helpers.DeployOperator(t, ctx)
+	helpers.AssertNoError(t, err)
+	t.Log("deployed operaot for addomn")
+
+	t.Run("addon-install", testsuites.ValidateAddonInstall)
+	//t.Run("addon-delete", testsuites.ValidateAddonDeletion)
```
Comment (sthaha) [inline test/testgroups/addon_crd.go]: delete this? 

Aren't we validating deletion?

File: test/helpers/helpers.go [hunk 22]
```diff
@@ -47,14 +48,14 @@ func WaitForDeploymentDeletion(t *testing.T, namespace, name string) error {
 	return err
 }
 
-func WaitForClusterCR(t *testing.T, name string) *op.Config {
+func WaitForClusterCR(t *testing.T, name string, obj runtime.Object) {
 	t.Helper()
 
 	objKey := types.NamespacedName{Name: name}
-	cr := &op.Config{}
+	//cr := &op.Config{}
```
Comment (sthaha) [inline test/helpers/helpers.go]: delete: commented code ?

File: deploy/resources/addons/dashboard/v0.1.0/gcr-tekton-dashboard.yaml (hunk 12 above)
Comment (akihikokuroda) [reply deploy/resources/addons/dashboard/v0.1.0/gcr-tekton-dashboard.yaml]: Yes, the v0.1.0 is not necessary

File: deploy/role.yaml (hunk 13 above)
Comment (akihikokuroda) [reply deploy/role.yaml]: Yes, it can.

File: deploy/role.yaml (hunk 14 above)
Comment (akihikokuroda) [reply deploy/role.yaml]: These are to avoid the following error:

{"level":"error","ts":1568832561.2466843,"logger":"ctrl.addon.addon install","msg":"failed to apply release.yaml","Request.Namespace":"","Request.NamespaceName":"/dashboard","Request.Name":"dashboard","error":"clusterroles.rbac.authorization.k8s.io \"tekton-dashboard-minimal\\
" is forbidden: attempt to grant extra privileges:

[{[use] [security.openshift.io] [securitycontextconstraints] [] []}
{[get] [extensions] [ingresses] [] []} {[list] [extensions] [ingresses] [] []}
{[watch] [extensions] [ingresses] [] []}
{[get] [apps] [ingresses] [] []}
{[list] [apps] [ingresses] [] []}
{[watch] [apps] [ingresses] [] []}
{[get] [dashboard.tekton.dev] [extensions] [] []}
{[list] [dashboard.tekton.dev] [extensions] [] []}
{[create] [dashboard.tekton.dev] [extensions] [] []}
{[update] [dashboard.tekton.dev] [extensions] [] []}
{[delete] [dashboard.tekton.dev] [extensions] [] []}
{[patch] [dashboard.tekton.dev] [extensions] [] []}
{[watch] [dashboard.tekton.dev] [extensions] [] []}]

File: deploy/role.yaml (hunk 15 above)
Comment (akihikokuroda) [reply deploy/role.yaml]: These are to avoid the following error:

{"level":"error","ts":1568832561.2466843,"logger":"ctrl.addon.addon install","msg":"failed to apply release.yaml","Request.Namespace":"","Request.NamespaceName":"/dashboard","Request.Name":"dashboard","error":"clusterroles.rbac.authorization.k8s.io \"tekton-dashboard-minimal\\
" is forbidden: attempt to grant extra privileges:

[{[use] [security.openshift.io] [securitycontextconstraints] [] []}
{[get] [extensions] [ingresses] [] []} {[list] [extensions] [ingresses] [] []}
{[watch] [extensions] [ingresses] [] []}
{[get] [apps] [ingresses] [] []}
{[list] [apps] [ingresses] [] []}
{[watch] [apps] [ingresses] [] []}
{[get] [dashboard.tekton.dev] [extensions] [] []}
{[list] [dashboard.tekton.dev] [extensions] [] []}
{[create] [dashboard.tekton.dev] [extensions] [] []}
{[update] [dashboard.tekton.dev] [extensions] [] []}
{[delete] [dashboard.tekton.dev] [extensions] [] []}
{[patch] [dashboard.tekton.dev] [extensions] [] []}
{[watch] [dashboard.tekton.dev] [extensions] [] []}]

File: pkg/controller/addon/addon_controller.go (hunk 16 above)
Comment (akihikokuroda) [reply pkg/controller/addon/addon_controller.go]: OK

File: pkg/controller/addon/addon_controller.go (hunk 17 above)
Comment (akihikokuroda) [reply pkg/controller/addon/addon_controller.go]: yes.

File: pkg/controller/addon/addon_controller.go (hunk 18 above)
Comment (akihikokuroda) [reply pkg/controller/addon/addon_controller.go]: Yes.

File: test/helpers/helpers.go (hunk 20 above)
Comment (akihikokuroda) [reply test/helpers/helpers.go]: yes

File: test/helpers/helpers.go (hunk 22 above)
Comment (akihikokuroda) [reply test/helpers/helpers.go]: yes

File: pkg/controller/addon/addon_controller.go (hunk 19 above)
Comment (akihikokuroda) [reply pkg/controller/addon/addon_controller.go]: ok

File: pkg/controller/addon/addon_controller.go [hunk 23]
```diff
@@ -0,0 +1,358 @@
...
+
+func (r *ReconcileAddon) refreshCR(res *op.Addon) error {
+	objKey := types.NamespacedName{
+		Namespace: res.Namespace,
+		Name:      res.Name,
+	}
+	return r.client.Get(context.TODO(), objKey, res)
+}
+
+func (r *ReconcileAddon) getPipelineRes() (*op.Config, error) {
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply pkg/controller/addon/addon_controller.go]: OK

**Comment by kaczyns** [comment]
Hello - I was wondering if this PR is still being considered or is being actively worked.  Over in https://github.com/kabanero-io/kabanero-operator/issues/141 we are considering moving the deploy of the Tekton dashboard and webhook extension from our install script, into the Kabanero operator, to solve an ordering issue.  Having the Tekton operator able to at least deploy the dashboard extension would be a real help.

**Comment by akihikokuroda** [comment]
@kaczyns I'm working on this.  @nikhil-thomas I'll create a PR for your branch soon.

File: test/testgroups/addon_crd.go (hunk 21 above)
Comment (akihikokuroda) [reply test/testgroups/addon_crd.go]: I add 2 delete validations.  delete addon CR and delete config CR.

File: deploy/resources/addons/dashboard/v0.1.0/gcr-tekton-dashboard.yaml (hunk 12 above)
Comment (akihikokuroda) [reply deploy/resources/addons/dashboard/v0.1.0/gcr-tekton-dashboard.yaml]: I remove webhooks-extension_release.yaml from v0.1.1.  The webhook-extension should be a separate addon.

**Comment by akihikokuroda** [comment]
Please review https://github.com/nikhil-thomas/operator/pull/2

**Comment by nikhil-thomas** [comment]
@akihikokuroda i shall address the comments and submit the changes within a day.

File: pkg/controller/addon/addon_controller.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (nikhil-thomas) [reply pkg/controller/addon/addon_controller.go]: `getPipelineRes` give more context. `getClusterConfi` can be confusing. Someone reading this can think ti has something to do with `kubeconfig`.

File: pkg/controller/addon/addon_controller.go (hunk 19 above)
Comment (nikhil-thomas) [reply pkg/controller/addon/addon_controller.go]: i prefer to use the  word `pipeline` in this context. We might know what we mean by `cluster - config` here. But it is a phrase widely used in Kubernetes ecosystem to denote something related to the cluster itself.

File: test/testsuites/addon_install_pipeline.go [hunk 24]
```diff
@@ -0,0 +1,277 @@
...
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name: "dashboard",
+		},
+		Spec: v1alpha1.AddonSpec{
+			Version: "v0.1.1",
+		},
+	}
+
+	cleanupOpetions := &test.CleanupOptions{
```
Comment (sthaha) [inline test/testsuites/addon_install_pipeline.go]: nit: typo: Options
